api/v1alpha1/backendtrafficpolicy_types.go

```diff
@@ -98,6 +103,15 @@ type BackendTrafficPolicySpec struct {
 	Compression []*Compression `json:"compression,omitempty"`
 }
 
+type HTTPOptions struct {
+	// UseClientProtocol configures Envoy to prefer sending requests to TLS enabled backends using
+	// the same HTTP protocol that the incoming request used. Defaults to false, which means
```

Comment:
prefer if we added a line in here saying something along the lines of - `this is especially useful when the client (connecting to the gateway) and backend support multiple HTTP versions (e.g. http/1.1 and http/2), and users would like to force the backend to use the same version the client is using`

Reply:
OK. Effectively this means that the backend must be TLS enabled, or it can't support multiple HTTP versions.

Reply:
can you elaborate on how TLS / ALPN is related here ? 
as far as in understand, this knob is proactively deciding the upstream protocol version to use based on downstream connection

Reply:
A backend without TLS will either support HTTP/1.1 or HTTP/2, but never both.  That's the point of `h2c` - it requires pre-knowledge on the part of the client that the server requires HTTP/2.

Since HTTP/2 is a binary protocol, if a client like Envoy Proxy starts a connection by sending a textual HTTP/1.1 request, the server will disconnect it. 
Since HTTP/1.1 is a textual protocol, if a client like Envoy Proxy starts a connection by sending an HTTP/2 request, the server will disconnect it.
Without TLS + ALPN, there is no way for a client like Envoy Proxy to negotiate with the server which protocol it can use, and this means that effectively upstream can't support _multiple_ versions of HTTP without TLS+ALPN.

> (like h2c to upstream, if downstream is h2c or h2)

We can configure the Envoy Proxy client to try to connect to upstream with the downstream protocol even without TLS on the upstream protocol. What it would do is automatically fail all requests where the downstream protocol doesn't match the single upstream protocol that is possible without upstream TLS.